ligo/haskell/src/Ligo/Typescript.hs

```diff
@@ -0,0 +1,440 @@
+      Nothing -> lookupEntryPointsAndTypes f1
+    rightEntrypoint = case extractAnn fa2 of
+      Just ann_ -> (Entrypoint ann_ (toEpType f2)) : (lookupEntryPointsAndTypes f2)
+      Nothing -> lookupEntryPointsAndTypes f2
+  in leftEntrypoint <> rightEntrypoint
+lookupEntryPointsAndTypes _ = []
+
+-- Check if the annotation in this annotation is among the
+-- ones we are not interested in, or empty. If so return a Nothing. Else
+-- wrap the annotation in a Just before returning it.
+extractAnn :: Annotation tag -> Maybe Text
+extractAnn ann_ =
+  case (ann_ == noAnn) || (isJust $ find (== ann_) ignoreEntypoints) of
+    True -> Nothing
+    _ -> Just $ unAnnotation ann_
+
+-- Some entrypoints we don't want to call directly.
+ignoreEntypoints :: [Annotation tag]
+ignoreEntypoints = ann <$>
+  [ "running"
+  , "xtzForbidden"
+  , "call_FA2"
+  , "migratable"
+  ]
+
+type Typename = Text
+type Fieldname = Text
+
+-- Represents a Typescript declaration. For now only contains one that
+-- defines a Type.
+data TsDecl
+  = TsType Typename TsTypeDef
+  deriving stock Generic
+
+-- Represents an instance of a Typescript type.
+data TsType
+  = TsNumber
+  | TsString
+  | TsBool
+  | TsUnit
+  | TsOption TsType
+  | TsList TsType
+  | TsMap TsType TsType
+  | TsBigMap TsType TsType
+  | TsCustom Typename
+  | TsStrLiteral Text
+  deriving stock Generic
+
+-- Implements the conversion of Type to its source code representation.
+instance Buildable TsType where
+  build = \case
+    TsNumber -> "number"
+    TsString -> "string"
+    TsBool -> "boolean"
+    TsUnit -> "{}"
+    TsOption t_ -> "null | " +| (build t_)
+    TsList t_ -> "Array<" +| t_ |+ ">"
+    TsMap k v -> "Map<" +| k |+ "," +| v |+ ">"
+    TsBigMap k v -> "Map<" +| k |+ "," +| v |+ ">"
+    TsCustom tn -> build tn
+    TsStrLiteral tn -> build $ "\""  <> tn <> "\""
+
+-- Represents a Typescript module with just enough stuff we are interested.
+data TsModule = TsModule
+  { tsmEpName :: Text
+  , tsmImports :: [Text]
+  , tsmDecls :: [TsDecl]
+  , tsmExports :: [Typename]
+  } deriving stock Generic
+
+type IsOptional = Bool
+
+-- Represents the right side of a Type definition in Typescript.
+data TsTypeDef
+  = TsUnion [TsType]
+  | TsInterface [(Fieldname, (TsType, IsOptional))]
+  | TsTuple [TsType]
+  | TsAlias TsType
+  deriving stock Generic
+
+-- Implements the conversion to source code.
+instance Buildable TsTypeDef where
+  build (TsUnion xs) = Fmt.indentF 2 $ (Fmt.blockListF' "|" build xs)
+  build (TsTuple fs) =  Fmt.unwordsF ["[", T.intercalate ", " (Fmt.pretty <$> fs), "];"]
+  build (TsInterface fns) =  Fmt.unlinesF ["{", Fmt.indentF 2 $ Fmt.unlinesF $ mkPair <$> fns, "};"]
+    where
+      mkPair :: (Fieldname, (TsType, IsOptional)) -> Text
+      mkPair (fn, (tn, isOptional)) =
+        case isOptional of
+          True -> fn <> "?: " <> (Fmt.pretty tn) <> ";"
+          False -> fn <> ": " <> (Fmt.pretty tn) <> ";"
+  build (TsAlias t) = build t <> ";"
```

Comment:
What version of TypeScript are we targetting? AFAIK newest versions don't require semicolons because of ASI (Automatic Semicolon Insertion), but I don't mind them either.

Reply:
Yes, IIRC semicolons were optional in Typescript from the begining. Not sure though. I just included it here since including them appeared to be more common in Typescript world.

Reply:
Ok, it doesn't return leaving them. I think there are even some cases that leaving them out doesn't do what you expect, for example:
```typescript
return
{
    foo: bar
}
```
Will become:
``` typescript
return;
{
    foo: bar
};
```
Unless the brackets were in the same line as the `return`.